File: ios/FluentAdFlowWidget/FluentAdFlowWidgetViewManager.m
//
//  FluentAdFlowWidgetViewManager.m
//  FluentAdFlowWidget
//
//  Created by Daniel Larner on 9/19/23.
//

#import <React/RCTViewManager.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(FluentAdFlowViewManager, RCTViewManager)
    RCT_EXPORT_VIEW_PROPERTY(email, NSString)
    RCT_EXPORT_VIEW_PROPERTY(firstName, NSString)
    RCT_EXPORT_VIEW_PROPERTY(lastName, NSString)
    RCT_EXPORT_VIEW_PROPERTY(orderId, NSString)
    RCT_EXPORT_VIEW_PROPERTY(transactionValue, NSString)
    RCT_EXPORT_VIEW_PROPERTY(zip, NSString)
@end

@interface RCT_EXTERN_MODULE(AdsViewEventEmitter, RCTEventEmitter)
    RCT_EXTERN_METHOD(supportedEvents)
    RCT_EXTERN_METHOD(requiresMainQueueSetup)
@end

File: ios/FluentAdFlowWidget/FluentAdFlowWidgetViewManager.swift
//
//  FluentAdFlowWidgetViewManager.swift
//  FluentAdFlowWidget
//
//  Created by Daniel Larner on 9/19/23.
//

import FluentAdFlowWidget

@objc(FluentAdFlowWidgetViewManager)
class FluentAdFlowWidgetViewManager: RCTViewManager {

  override func view() -> (FluentAdFlowWidgetView) {
    return FluentAdFlowWidgetView()
  }

  @objc override static func requiresMainQueueSetup() -> Bool {
    return false
  }
}

class FluentAdFlowWidgetView : FluentAdFlowWidgetView  {
  required init(frame: CGRect){
    super.init(frame: frame)
    super.events.listenTo(eventName: "onHide", action: onHidden)
    super.events.listenTo(eventName: "onShow", action: onShow)
  }

  required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
  }

  override func didSetProps(_ changedProps: [String]!) {
    super.updateView(changedProps)
  }

  func onHidden(data: Any?) {
    print("-------- on Hidden -------------")
    AdFlowViewEventEmitter.emitter.sendEvent(withName: "onHide", body: ["hidden"])
  }

  func onShow(data: Any?) {
    print("-------- on Show -------------")
    AdFlowViewEventEmitter.emitter.sendEvent(withName: "onShow", body: ["show"])
  }
}


@objc(AdFlowViewEventEmitter)
open class AdFlowViewEventEmitter: RCTEventEmitter {

  public static var emitter: AdFlowViewEventEmitter!

  override init() {
    super.init()
      AdFlowViewEventEmitter.emitter = self
  }

  @objc
  override public func supportedEvents() -> [String] {
    ["onReady", "onPending", "onFailure", "onHide", "onShow"]
  }

  @objc override public static func requiresMainQueueSetup() -> Bool {
        return false
    }
}
